reactioncommerce/reaction
# Finish simple-pricing plugin

Resolves #4969 
Resolves #4970   
Impact: **major**  
Type: **feature**

## Big Picture Benefit
Because all pricing logic and schemas are encapsulated in the included `simple-pricing` plugin, installations that have their own custom pricing plugin do not have to jump through hoops to keep core pricing code happy. Default pricing behavior can be disabled by disabling the `simple-pricing` plugin.

## Changes
- Removed the deprecated `price` field from `CatalogProduct` and `CatalogProductVariant`. Ensured that no code uses it.  I opted not to do a migration to remove this from the database. It will be removed as each product is published, or you may manually unset it.
- Moved price-related functions, code, and schemas into the `simple-pricing` plugin. This plugin was created prior to this for this purpose, but we never finished migrating all pricing code to there.
- Some significant improvements to database lookups for pricing-related fields.

NOTE: Some operator UI client code has not yet been moved into the simple-pricing plugin. This will be done in a future PR.

## Breaking changes
The deprecated `price` field has been removed from `CatalogProduct` and `CatalogProductVariant`.

## Tangential Changes
Because it was needed for this work, I've added `app.registerPlugin`, which is a version of `Reaction.registerPackage` that can be called in a pure Node app. Soon we will update all plugins to use this, but for now `Reaction.registerPackage` still works and the API is the same for both functions.

## Testing
- Verify that prices look correct in UI everywhere
- Verify operator can update prices and publish updated prices to catalog
- Verify you can check out on storefront with correct prices
- Verify that you can disable the `simple-pricing` plugin, add a custom pricing plugin to replace it, and everything related to prices continues to work.
